From 36a32537a1a0044b474e63603881f5b96b3e7fff Mon Sep 17 00:00:00 2001
From: Peter Rifel <pgrifel@gmail.com>
Date: Mon, 29 Jul 2024 21:34:38 -0500
Subject: [PATCH] Fix InTreePluginAWSUnregister feature gate removal in 1.31

Cherry-pick of https://github.com/kubernetes/kops/commit/36a32537a1a0044b474e63603881f5b96b3e7fff

This patch can be removed when this commit is part of a kOps release.
---
 pkg/model/components/apiserver.go             |  2 +-
 pkg/model/components/kubecontrollermanager.go |  2 +-
 pkg/model/components/kubelet.go               |  8 ++--
 pkg/model/components/kubescheduler.go         |  2 +-
 4 files changed, 7 insertions(+), 7 deletions(-)

diff --git a/pkg/model/components/apiserver.go b/pkg/model/components/apiserver.go
index 60a8fb5146..a0f17052d1 100644
--- a/pkg/model/components/apiserver.go
+++ b/pkg/model/components/apiserver.go
@@ -183,7 +183,7 @@ func (b *KubeAPIServerOptionsBuilder) BuildOptions(o interface{}) error {
 
 	if clusterSpec.CloudProvider.AWS != nil {
 
-		if _, found := c.FeatureGates["InTreePluginAWSUnregister"]; !found {
+		if _, found := c.FeatureGates["InTreePluginAWSUnregister"]; !found && b.IsKubernetesLT("1.31") {
 			c.FeatureGates["InTreePluginAWSUnregister"] = "true"
 		}
 
diff --git a/pkg/model/components/kubecontrollermanager.go b/pkg/model/components/kubecontrollermanager.go
index cceb562d7c..5d142c9f3e 100644
--- a/pkg/model/components/kubecontrollermanager.go
+++ b/pkg/model/components/kubecontrollermanager.go
@@ -149,7 +149,7 @@ func (b *KubeControllerManagerOptionsBuilder) BuildOptions(o interface{}) error
 			kcm.FeatureGates = make(map[string]string)
 		}
 
-		if _, found := kcm.FeatureGates["InTreePluginAWSUnregister"]; !found {
+		if _, found := kcm.FeatureGates["InTreePluginAWSUnregister"]; !found && b.IsKubernetesLT("1.31") {
 			kcm.FeatureGates["InTreePluginAWSUnregister"] = "true"
 		}
 
diff --git a/pkg/model/components/kubelet.go b/pkg/model/components/kubelet.go
index 8aeb6a59b7..e1b55031f7 100644
--- a/pkg/model/components/kubelet.go
+++ b/pkg/model/components/kubelet.go
@@ -172,12 +172,12 @@ func (b *KubeletOptionsBuilder) BuildOptions(o interface{}) error {
 	}
 
 	if clusterSpec.CloudProvider.AWS != nil {
-		if _, found := clusterSpec.Kubelet.FeatureGates["CSIMigrationAWS"]; !found && b.IsKubernetesLT("1.27") {
-			clusterSpec.Kubelet.FeatureGates["CSIMigrationAWS"] = "true"
+		if _, found := clusterSpec.Kubelet.FeatureGates["InTreePluginAWSUnregister"]; !found && b.IsKubernetesLT("1.31") {
+			clusterSpec.Kubelet.FeatureGates["InTreePluginAWSUnregister"] = "true"
 		}
 
-		if _, found := clusterSpec.Kubelet.FeatureGates["InTreePluginAWSUnregister"]; !found {
-			clusterSpec.Kubelet.FeatureGates["InTreePluginAWSUnregister"] = "true"
+		if _, found := clusterSpec.Kubelet.FeatureGates["CSIMigrationAWS"]; !found && b.IsKubernetesLT("1.27") {
+			clusterSpec.Kubelet.FeatureGates["CSIMigrationAWS"] = "true"
 		}
 	}
 
diff --git a/pkg/model/components/kubescheduler.go b/pkg/model/components/kubescheduler.go
index ff50a001aa..dab0591d4b 100644
--- a/pkg/model/components/kubescheduler.go
+++ b/pkg/model/components/kubescheduler.go
@@ -63,7 +63,7 @@ func (b *KubeSchedulerOptionsBuilder) BuildOptions(o interface{}) error {
 			config.FeatureGates = make(map[string]string)
 		}
 
-		if _, found := config.FeatureGates["InTreePluginAWSUnregister"]; !found {
+		if _, found := config.FeatureGates["InTreePluginAWSUnregister"]; !found && b.IsKubernetesLT("1.31") {
 			config.FeatureGates["InTreePluginAWSUnregister"] = "true"
 		}
 
